### PR TITLE
add timeout to remote schema calls (close #2501)

### LINKED ISSUE #2501: Increase response timeout for remote schemas
We currently use the hasura graphql-engine with a custom remote schema that implements certain special graphql mutations. Some of these mutations can run longer than 30 seconds, which seems to be the default http request timeout of hasura. After the 30 seconds we receive the following error:

```
{
  "errors": [
    {
      "extensions": {
        "path": "$",
        "code": "unexpected"
      },
      "message": "HttpExceptionRequest Request {\n  host                 = \"our-host\"\n  port                 = 80\n  secure               = False\n  requestHeaders       = [(\"Content-Type\",\"application/json\"),(\"User-Agent\",\"hasura-graphql-engine/v1.0.0-beta.2\"),(\"x-hasura-role\",\"admin\")]\n  path                 = \"/graphql\"\n  queryString          = \"\"\n  method               = \"POST\"\n  proxy                = Nothing\n  rawBody              = False\n  redirectCount        = 10\n  responseTimeout      = ResponseTimeoutDefault\n  requestVersion       = HTTP/1.1\n}\n ResponseTimeout"
    }
  ]
}
```

It would be nice if the response timeout for remote schemas would be configurable or at least could be increased to something higher than 30 seconds, which seems pretty short to us.

### PR DESCRIPTION
**Pull request by tirumaraiselvan**
### Description

 Adds an optional `timeout_seconds` argument to `add_remote_schema` API

### Affected components 

- Server
- Console
- Docs

### Related Issues

#2501

### CODE REVIEW

**Comment by rikinsk** [comment]
@tirumaraiselvan Can you update the API changes in docs

File: server/src-lib/Hasura/Server/Utils.hs [hunk 1]
```diff
@@ -212,7 +212,7 @@ httpExceptToJSON e = case e of
 -- ignore the following request headers from the client
 commonClientHeadersIgnored :: (IsString a) => [a]
 commonClientHeadersIgnored =
-  [ "Content-Length", "Content-MD5", "User-Agent", "Host"
+  [ "Content-Length", "Content-MD5", "Host"
```
Comment (0x777) [inline server/src-lib/Hasura/Server/Utils.hs]: Why is this removed?

Comment (tirumaraiselvan) [reply server/src-lib/Hasura/Server/Utils.hs]: https://github.com/hasura/graphql-engine/issues/2572

Comment (tirumaraiselvan) [reply server/src-lib/Hasura/Server/Utils.hs]: I am not sure if this was the best way. Review requested.

File: server/src-lib/Hasura/RQL/Types/RemoteSchema.hs [hunk 2]
```diff
@@ -30,6 +30,7 @@ data RemoteSchemaInfo
   { rsUrl              :: !N.URI
   , rsHeaders          :: ![HeaderConf]
   , rsFwdClientHeaders :: !Bool
+  , rsTimeoutSeconds   :: !Int
```
Comment (lexi-lambda) [inline server/src-lib/Hasura/RQL/Types/RemoteSchema.hs]: Consider maybe using [`DiffTime`](https://hackage.haskell.org/package/time-1.9.3/docs/Data-Time-Clock.html#t:DiffTime) here to add a small amount of type safety? The main advantage, imo, is that the units are unambiguous (a `DiffTime` is always treated to mean seconds when converted to/from other numeric types).

File: server/src-lib/Hasura/HTTP.hs [hunk 3]
```diff
@@ -25,9 +26,12 @@ hdrsToText hdrs =
 wreqOptions :: HTTP.Manager -> [HTTP.Header] -> Wreq.Options
 wreqOptions manager hdrs =
   Wreq.defaults
-  & Wreq.headers .~  contentType : userAgent : hdrs
+  & Wreq.headers .~  defaultHeaders <> hdrs
```
[the remark this answers is not in the document]
Comment (tirumaraiselvan) [reply server/src-lib/Hasura/HTTP.hs]: Right! Thanks.

File: server/src-lib/Hasura/Server/Utils.hs (hunk 1 above)
Comment (0x777) [reply server/src-lib/Hasura/Server/Utils.hs]: I don't think we can forward the original `user-agent` as is given that we are not a proxy (currently we kind of are but we won't be in the future when you have remote joins). Maybe we can add headers like `hge-client-user-agent` and `hge-client-ip`?

Comment (tirumaraiselvan) [reply server/src-lib/Hasura/Server/Utils.hs]: Okay, I am just going to revert this commit. And take it in a separate PR

File: server/src-lib/Hasura/RQL/Types/RemoteSchema.hs (hunk 2 above)
Comment (tirumaraiselvan) [reply server/src-lib/Hasura/RQL/Types/RemoteSchema.hs]: Turns out `DiffTime` is not an instance of `Lift` which is creeping into 20 odd places.

File: server/src-lib/Hasura/Events/Lib.hs [hunk 4]
```diff
@@ -345,10 +345,16 @@ decodeHeader logenv headerInfos (hdrName, hdrVal)
      decodeBS = TE.decodeUtf8With TE.lenientDecode
 
 addDefaultHeaders :: [HTTP.Header] -> [HTTP.Header]
-addDefaultHeaders hdrs = hdrs ++
-  [ (CI.mk "Content-Type", "application/json")
-  , (CI.mk "User-Agent", "hasura-graphql-engine/" <> T.encodeUtf8 currentVersion)
-  ]
+addDefaultHeaders hdrs = defaultHeaders <> rmDefaultHeaders hdrs
```
Comment (tirumaraiselvan) [inline server/src-lib/Hasura/Events/Lib.hs]: @lexi-lambda This change has led me to believe that we have HTTP clients which are not using the Hasura.HTTP library basically (event trigger module and remote schema module):

The only modules which seem to be using this are:

Telemetry
Auth
Auth/JWT
CheckUpdates

Is it fine for EventTriggers and RemoteSchemas to maintain their own clients because these might be specific to these modules.

File: server/src-lib/Hasura/RQL/Types/RemoteSchema.hs (hunk 2 above)
Comment (lexi-lambda) [reply server/src-lib/Hasura/RQL/Types/RemoteSchema.hs]: Odd. I guess that’s alright, then, it’s not a big deal.

File: server/src-lib/Hasura/Events/Lib.hs (hunk 4 above)
Comment (lexi-lambda) [reply server/src-lib/Hasura/Events/Lib.hs]: This function appears to prioritize default headers over the headers that were passed in, but the code in `GraphQL.Execute` appears to prioritize the headers that were passed in over the default headers. Is the difference intentional?

Also, can this code at least share the `defaultHeaders` declaration with the one in `Hasura.HTTP`? Having it duplicated in both places seems worth avoiding if we can help it.

Comment (tirumaraiselvan) [reply server/src-lib/Hasura/Events/Lib.hs]: This was my question. Is it worth sharing these tiny functions, when the bulk of the client is specified per module itself? Shouldn't HTTP.Hasura give out a client with these baked in for standard usage? Those who are not using it may do their own thing including defaultHeaders?

Comment (lexi-lambda) [reply server/src-lib/Hasura/Events/Lib.hs]: I don’t know, as I haven’t looked at the code in detail, but my gut instinct is that

  1. these modules should probably be using `Hasura.HTTP`,

  2. but if they aren’t, and you don’t want to do the refactoring to make them use it right now, sharing the default headers is an alright first step.

If the first assumption isn’t true, then you could be right, but I’d be surprised if it were. Is there a reason these modules shouldn’t be using the same library?

Comment (tirumaraiselvan) [reply server/src-lib/Hasura/Events/Lib.hs]: I don't think there is any real reason except the library didn't have proper functions to use with Event Triggers. I guess we should make an attempt to consolidate the library then.

Cool, I am just going to share the functions then. Let's do the full consolidation in a separate no-op refactor PR.

Comment (lexi-lambda) [reply server/src-lib/Hasura/Events/Lib.hs]: Sounds good. 👍
